File: db/ts/src/sanitize.ts
import {
  hashEmail,
  maskIpSubnet,
  hashUserAgent,
  hashDeviceId,
} from './crypto';

const SECRET_KEY_PATTERN = /^(key_secret|razorpay_key_secret|secret|password|authorization|private_key|api_key)$/i;
const SAFE_TOKEN_PATTERN = /(_hash|_preview)$/i;
const EMAIL_KEY_PATTERN = /email/i;
const IP_KEY_PATTERN = /^(ip|client_ip|ip_address|remote_addr)$/i;
const UA_KEY_PATTERN = /user_agent|useragent/i;
const DEVICE_KEY_PATTERN = /device_id|deviceid/i;

const IP_VALUE_PATTERN = /^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/;
const EMAIL_VALUE_PATTERN = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

/**
 * Recursively sanitize audit payload data:
 * - Redacts raw merchant secrets / passwords / tokens
 * - Converts emails to SHA-256 hashes
 * - Truncates IP addresses to /24 subnets
 * - Hashes User-Agent and Device ID strings
 * (D-07, PRD §15.3)
 */
export function sanitizeAuditData(data: Record<string, any>): Record<string, any> {
  if (data === null || typeof data !== 'object') {
    return data;
  }

  if (Array.isArray(data)) {
    return data.map((item) =>
      typeof item === 'object' && item !== null ? sanitizeAuditData(item) : item
    ) as any;
  }

  const sanitized: Record<string, any> = {};

  for (const [key, value] of Object.entries(data)) {
    if (value === null || value === undefined) {
      sanitized[key] = value;
      continue;
    }

    // 1. Redact secrets
    if (SECRET_KEY_PATTERN.test(key) && !SAFE_TOKEN_PATTERN.test(key)) {
      sanitized[key] = '[REDACTED]';
      continue;
    }

    if (key.toLowerCase() === 'maas_token' && !SAFE_TOKEN_PATTERN.test(key)) {
      sanitized[key] = '[REDACTED]';
      continue;
    }

    // 2. Objects and Arrays recurse
    if (typeof value === 'object') {
      sanitized[key] = sanitizeAuditData(value);
      continue;
    }

    if (typeof value === 'string') {
      // 3. Email sanitization
      if (EMAIL_KEY_PATTERN.test(key) && !SAFE_TOKEN_PATTERN.test(key)) {
        sanitized[key] = hashEmail(value);
        continue;
      }
      if (EMAIL_VALUE_PATTERN.test(value.trim())) {
        sanitized[key] = hashEmail(value);
        continue;
      }

      // 4. IP sanitization
      if (IP_KEY_PATTERN.test(key)) {
        sanitized[key] = maskIpSubnet(value);
        continue;
      }
      if (IP_VALUE_PATTERN.test(value.trim())) {
        sanitized[key] = maskIpSubnet(value);
        continue;
      }

      // 5. User-Agent sanitization
      if (UA_KEY_PATTERN.test(key) && !SAFE_TOKEN_PATTERN.test(key)) {
        sanitized[key] = hashUserAgent(value);
        continue;
      }

      // 6. Device ID sanitization
      if (DEVICE_KEY_PATTERN.test(key) && !SAFE_TOKEN_PATTERN.test(key)) {
        sanitized[key] = hashDeviceId(value);
        continue;
      }
    }

    sanitized[key] = value;
  }

  return sanitized;
}

File: db/py/nexus_db/sanitize.py
import re
from typing import Any
from .crypto import (
    hash_email,
    mask_ip_subnet,
    hash_user_agent,
    hash_device_id,
)

SECRET_KEY_PATTERN = re.compile(
    r"^(key_secret|razorpay_key_secret|secret|password|authorization|private_key|api_key)$",
    re.IGNORECASE,
)
SAFE_TOKEN_PATTERN = re.compile(r"(_hash|_preview)$", re.IGNORECASE)
EMAIL_KEY_PATTERN = re.compile(r"email", re.IGNORECASE)
IP_KEY_PATTERN = re.compile(r"^(ip|client_ip|ip_address|remote_addr)$", re.IGNORECASE)
UA_KEY_PATTERN = re.compile(r"user_agent|useragent", re.IGNORECASE)
DEVICE_KEY_PATTERN = re.compile(r"device_id|deviceid", re.IGNORECASE)

IP_VALUE_PATTERN = re.compile(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$")
EMAIL_VALUE_PATTERN = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")


def sanitize_audit_data(data: Any) -> Any:
    """Recursively sanitize audit payload data:

    - Redacts raw merchant secrets / passwords / tokens
    - Converts emails to SHA-256 hashes
    - Truncates IP addresses to /24 subnets
    - Hashes User-Agent and Device ID strings
    (D-07, PRD §15.3)
    """
    if data is None:
        return None

    if isinstance(data, list):
        return [sanitize_audit_data(item) for item in data]

    if not isinstance(data, dict):
        return data

    sanitized: dict[str, Any] = {}

    for key, value in data.items():
        if value is None:
            sanitized[key] = None
            continue

        # 1. Redact secrets
        if SECRET_KEY_PATTERN.match(key) and not SAFE_TOKEN_PATTERN.search(key):
            sanitized[key] = "[REDACTED]"
            continue

        if key.lower() == "maas_token" and not SAFE_TOKEN_PATTERN.search(key):
            sanitized[key] = "[REDACTED]"
            continue

        # 2. Dicts and Lists recurse
        if isinstance(value, (dict, list)):
            sanitized[key] = sanitize_audit_data(value)
            continue

        if isinstance(value, str):
            # 3. Email sanitization
            if EMAIL_KEY_PATTERN.search(key) and not SAFE_TOKEN_PATTERN.search(key):
                sanitized[key] = hash_email(value)
                continue
            if EMAIL_VALUE_PATTERN.match(value.strip()):
                sanitized[key] = hash_email(value)
                continue

            # 4. IP sanitization
            if IP_KEY_PATTERN.match(key):
                sanitized[key] = mask_ip_subnet(value)
                continue
            if IP_VALUE_PATTERN.match(value.strip()):
                sanitized[key] = mask_ip_subnet(value)
                continue

            # 5. User-Agent sanitization
            if UA_KEY_PATTERN.search(key) and not SAFE_TOKEN_PATTERN.search(key):
                sanitized[key] = hash_user_agent(value)
                continue

            # 6. Device ID sanitization
            if DEVICE_KEY_PATTERN.search(key) and not SAFE_TOKEN_PATTERN.search(key):
                sanitized[key] = hash_device_id(value)
                continue

        sanitized[key] = value

    return sanitized
